craft_application/application.py

```diff
@@ -583,7 +583,10 @@ def _transform_project_yaml(
         )
 
         # Perform variable expansion.
-        self._expand_environment(yaml_data)
+        self._expand_environment(
+            yaml_data=yaml_data,
+            build_for=build_for or util.get_host_architecture(),
```

Comment:
I think this fix is correct but insufficient. I can't comment on the line on this PR but [this bit of code](https://github.com/canonical/craft-application/blob/main/craft_application/application.py#L283) in `get_project()`:

```python
        self._full_build_plan = build_planner.get_build_plan()
        self._build_plan = filter_plan(
            self._full_build_plan, platform, build_for, host_arch
        )

        if platform and not build_for:
            all_platforms = {b.platform: b for b in self._full_build_plan}
            if platform not in all_platforms:
                raise errors.InvalidPlatformError(platform, list(all_platforms.keys()))
            build_for = all_platforms[platform].build_for
```
... I think this is right for the case where `--platform` is provided, but wrong in a different case. Consider this snippet from ondrak today:

```yaml
platforms:
    arm64:
        build-on: [amd64, arm64]
        build-for: arm64
```

... in this case, calling just `snapcraft pack --destructive-mode` _without_ `--platform` or `--build-for` would still result in `build_for` being `None`, which is incorrect because from the platform declaration it should clearly be `arm64`.
My suggestion is to add a bit like this at the end of that snippet:

```python
        if platform and not build_for:
            all_platforms = {b.platform: b for b in self._full_build_plan}
            if platform not in all_platforms:
                raise errors.InvalidPlatformError(platform, list(all_platforms.keys()))
            build_for = all_platforms[platform].build_for

        # new
        if build_for is None and self._build_plan:
            build_for = self._build_plan[0].build_for
```

This should be ok because destructive-mode builds will only run if the build plan has a single element, and in managed runs `build_for` won't be None
    
```

Reply:
Ooh, you are very correct!

I took this one step further by removing near-duplicate code in `_transform_project_yaml()`.

I would like to take this another step further by also moving the last resort `build_for = build_on` from `_transform_project_yaml()` to `get_project()`, but this will change the `build_for` that gets passed to `_extra_yaml_transforms()`.  It won't break our applications but it's too big of a change for a hotfix.